File: library/src/main/java/com/xwray/groupie/DiffTask.java
package com.xwray.groupie;

import android.os.AsyncTask;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;

import java.lang.ref.WeakReference;
import java.util.Collection;
import java.util.List;

/**
 * An async task implementation that runs {@link DiffUtil#calculateDiff(DiffUtil.Callback)}
 * in a background thread.
 * <p>This task is executed via {@link AsyncDiffUtil#calculateDiff(List, List, OnAsyncUpdateListener, boolean)}.
 */
class DiffTask extends AsyncTask<Void, Void, DiffUtil.DiffResult> {
    @NonNull
    private final DiffCallback diffCallback;
    private final WeakReference<AsyncDiffUtil> asyncListDiffer;
    private final int runGeneration;
    private final boolean detectMoves;
    @Nullable
    private WeakReference<OnAsyncUpdateListener> onAsyncUpdateListener;
    private Exception backgroundException = null;

    DiffTask(@NonNull AsyncDiffUtil asyncDiffUtil,
             @NonNull DiffCallback callback,
             int runGeneration,
             boolean detectMoves,
             @Nullable OnAsyncUpdateListener onAsyncUpdateListener) {
        this.diffCallback = callback;
        this.asyncListDiffer = new WeakReference<>(asyncDiffUtil);
        this.runGeneration = runGeneration;
        this.detectMoves = detectMoves;
        if (onAsyncUpdateListener != null) {
            this.onAsyncUpdateListener = new WeakReference<>(onAsyncUpdateListener);
        }
    }

    @Override
    @Nullable
    protected DiffUtil.DiffResult doInBackground(Void... voids) {
        try {
            return DiffUtil.calculateDiff(diffCallback, detectMoves);
        } catch (Exception e) {
            backgroundException = e;
            return null;
        }
    }

    @Override
    protected void onPostExecute(@Nullable DiffUtil.DiffResult diffResult) {
        if (backgroundException != null) {
            throw new RuntimeException(backgroundException);
        }
        AsyncDiffUtil async = asyncListDiffer.get();

        if (!shouldDispatchResult(diffResult, async)) {
            return;
        }

        AsyncDiffUtil.Callback asyncDiffUtilCallback = async.asyncDiffUtilCallback;
        asyncDiffUtilCallback.onDispatchAsyncResult(diffCallback.mergeGroups());

        diffResult.dispatchUpdatesTo(asyncDiffUtilCallback);
        if (onAsyncUpdateListener != null && onAsyncUpdateListener.get() != null) {
            onAsyncUpdateListener.get().onUpdateComplete();
        }
    }

    private boolean shouldDispatchResult(@Nullable DiffUtil.DiffResult diffResult, AsyncDiffUtil async) {
        return diffResult != null && async != null && runGeneration == async.getMaxScheduledGeneration();
    }
}

File: library-kotlin/src/main/java/com/wray/groupiekotlin/DiffTask.kt
package com.wray.groupiekotlin

import android.os.AsyncTask
import androidx.recyclerview.widget.DiffUtil
import java.lang.ref.WeakReference

/**
 * An async task implementation that runs [DiffUtil.calculateDiff]
 * in a background thread. This task will call [AsyncDiffUtil.Callback.onDispatchAsyncResult]
 * passing the new list just before dispatching the diff result to the provided
 * [DiffUtil.Callback] so that the new list.
 *
 * This task is executed via [AsyncDiffUtil.calculateDiff].
 */
internal class DiffTask(
    asyncDiffUtil: AsyncDiffUtil,
    private val diffCallback: DiffUtil.Callback,
    private val runGeneration: Int,
    private val detectMoves: Boolean,
    onAsyncUpdateListener: OnAsyncUpdateListener?
) : AsyncTask<Void, Void, DiffUtil.DiffResult>() {

    private val asyncListDiffer: WeakReference<AsyncDiffUtil> = WeakReference(asyncDiffUtil)
    private var onAsyncUpdateListener: WeakReference<OnAsyncUpdateListener>? = null
    private var backgroundException: Exception? = null

    init {
        if (onAsyncUpdateListener != null) {
            this.onAsyncUpdateListener = WeakReference(onAsyncUpdateListener)
        }
    }

    override fun doInBackground(vararg voids: Void): DiffUtil.DiffResult? =
        try {
            DiffUtil.calculateDiff(diffCallback, detectMoves)
        } catch (e: Exception) {
            backgroundException = e
            null
        }


    override fun onPostExecute(diffResult: DiffUtil.DiffResult?) {
        if (backgroundException != null) {
            throw RuntimeException(backgroundException)
        }
        val async = asyncListDiffer.get()
        if (shouldDispatchResult(diffResult, async)) {
            async?.asyncDiffUtilCallback?.let { callback ->
                callback.onDispatchAsyncResult(async.groups)
                diffResult?.dispatchUpdatesTo(callback)
            }

            onAsyncUpdateListener?.get()?.onUpdateComplete()
        }
    }

    private fun shouldDispatchResult(diffResult: DiffUtil.DiffResult?, async: AsyncDiffUtil?): Boolean =
        diffResult != null && async != null && runGeneration == async.maxScheduledGeneration

}
